### DISCUSSION TITLE
Revcloud

### DISCUSSION THREAD

**Discussion by gswifort** [mozman/ezdxf#1046] [Q&A] [answered]
Are there any plans to add REVCLOUD support?

**Comment by mozman** [comment]
No.

**Comment by mozman** [comment]
FYI: REVCLOUD is not a special DXF element, just a strange LWPOLYLINE with lots of bulges

![image](https://github.com/mozman/ezdxf/assets/4230966/a947b126-1d68-4afb-8b9f-ced07d28a6fd)

**Comment by mozman** [comment]
This script renders something similar, maybe I add the `revcloud()` function to the library:

```python
# Copyright (c) 2024, Manfred Moitzi
# License: MIT License
from typing import Sequence
from pathlib import Path
import math
import ezdxf
from ezdxf.math import UVec, Vec2

CWD = Path("~/Desktop/Outbox").expanduser()
if not CWD.exists():
    CWD = Path(".")


def revcloud(
    points: UVec,
    segment_length: float,
    bulge: float = 0.5,
    start_width: float = 0.0,
    end_width: float = 0.0,
) -> list[Sequence[float]]:
    vertices = Vec2.list(points)
    if not vertices[0].isclose(vertices[-1]):
        vertices.append(vertices[0])
    lw_points: list[Sequence[float]] = []

    for s, e in zip(vertices, vertices[1:]):
        lw_points.append((s.x, s.y, start_width, end_width, bulge))
        diff = e - s
        length = diff.magnitude
        if length <= segment_length:
            continue

        count = math.ceil(length / segment_length)
        _segment_length = length / count
        offset = diff.normalize(_segment_length)
        for _ in range(count):
            s += offset
            lw_points.append((s.x, s.y, start_width, end_width, bulge))
    return lw_points


def main():
    doc = ezdxf.new()
    msp = doc.modelspace()
    lw_points = revcloud(
        [(0, 0), (1, 0), (1, 1), (0, 1)],
        segment_length=0.1,
        bulge=0.5,
        start_width=0,
        end_width=0.01,
    )
    msp.add_lwpolyline(lw_points)
    doc.saveas(CWD / "revcloud.dxf")


if __name__ == "__main__":
    main()

```

![image](https://github.com/mozman/ezdxf/assets/4230966/2b5d2aaf-9975-4479-bad9-d8c387033dac)

**Reply by gswifort** [reply]
Thank you !
You can also add xdata so that Autocad sees it as revcloud.
```py
def revcloud(
    points: UVec,
    segment_length: float,
    bulge: float = 0.5,
    start_width: float = 0.0,
    end_width: float = 0.0,
) -> tuple[list[Sequence[float]], tuple]:
    """
    Examples:
        >>> doc = ezdxf.new()
        >>> msp = doc.modelspace()
        >>> lw_points, xdata_args = revcloud(
            [(0, 0), (1, 0), (1, 1), (0, 1)],
            segment_length=0.1,
            bulge=0.5,
            start_width=0,
            end_width=0.01,
        )
        >>> pline = msp.add_lwpolyline(lw_points)
        >>> pline.set_xdata(*xdata_args)
    """
    vertices = Vec2.list(points)
    if not vertices[0].isclose(vertices[-1]):
        vertices.append(vertices[0])
    lw_points: list[Sequence[float]] = []

    for s, e in zip(vertices, vertices[1:]):
        lw_points.append((s.x, s.y, start_width, end_width, bulge))
        diff = e - s
        length = diff.magnitude
        if length <= segment_length:
            continue

        count = math.ceil(length / segment_length)
        _segment_length = length / count
        offset = diff.normalize(_segment_length)
        for _ in range(count):
            s += offset
            lw_points.append((s.x, s.y, start_width, end_width, bulge))

    xdata_args = ("RevcloudProps", ((1001, "RevcloudProps"), (1070, 0), (1040, segment_length)))  # ← ← ←
    return lw_points, xdata_args
```

**Reply by mozman** [reply]
Don't forget to add `RevcloudProps`  to the APPID table, otherwise AutoCAD will not open the file. The XDATA tags do not require the (1001, "RevcloudProps") tag, that will be inserted automatically - but it doesn't harm either.

**Reply by gswifort** [reply]
Hello, just a few more comments:
- you should not add the final coordinates of a given section because they duplicate the first coordinate of the next section
```python
for _ in range(count-1):
    ...
```
- if you set bulge to a value other than AutoCAD's default value, it will not read the entity as Revcloud even if you add XData. Default value is `bulge=0.52056705` (0.52 is also accepted)

**Reply by mozman** [reply]
I change the default bulge value to 0.52... 

If the first vertex matches the last vertex, the polyline is explicitly closed. This is not a problem for any CAD application - if so, it is not a true CAD application. 😊

**Reply by gswifort** [reply]
I didn't mean the first and last vertex of an explicitly closed polyline, but the "middle" vertices, e.g. for [this example](https://github.com/mozman/ezdxf/blob/master/examples/render/revcloud.py) `lw_points[10] == lw_points[11]`, `lw_points[21] == lw_points[22]` etc.
This is still not a problem for CAD applications, I just wanted to point out that it is an unnecessary duplication.

**Reply by mozman** [reply]
Got it!

**Reply by gswifort** [reply]
You can also add a comment in the docstring that the vertices should be given counterclockwise, otherwise the bulge will be inward (which is not the intended effect).
Alternatively, you can estimate whether most of the figure has the outer edge on the left or on the right (I don't know if you already have a function for this) - then you just need to set the bulge with a minus sign

![Zrzut ekranu 2024-03-19 172007](https://github.com/mozman/ezdxf/assets/131259064/f54048ae-ed53-409f-923c-2f9a2888d906)

**Reply by mozman** [reply]
I have a function to test if a polygon is clockwise or counter-clockwise oriented.

I will add such a test to `revcloud.add_entity()` and negate the bulge value for clockwise oriented polygons, this is accepted by BricsCAD and DWG TrueView. The low-level function `revcloud.points()` remains unchanged - maybe someone likes this shape better.

**Comment by mozman** [comment, marked as answer]
Example script to add a true REVCLOUD object: https://github.com/mozman/ezdxf/blob/master/examples/revcloud.py

These conditions must be met to create a true REVCLOUD representation:

- bulge value is 0.52056705
- LWPOLYLINE closed flag must be set
- XDATA "RevcloudProps" must be present    

![image](https://github.com/mozman/ezdxf/assets/4230966/5ccc188e-1133-42e9-8e1d-a09a54c03981)

Whether the first vertex coincides with the last vertex or not is not important.

EDIT: moved module to `ezdxf.revcloud` as it can now do more than just render the required `LWPOLYLINE` points.
